chanbackup: archive old channel backup files

Abdulkbk opened the pull request:
Closes #8906 

## Change Description
From the [feature description](https://github.com/lightningnetwork/lnd/issues/8906):
> To ensure that certain classes of recovery are always possible, we should considering adding a mode to never delete channel backups, and only rename or move them to a special folder that stores archived channels. They're relatively small (~300 bytes or so), so it shouldn't take up a significant amount of disk space.


This PR allows for the archiving of channel backups, enabling users to choose whether to permanently delete previous backup files or archive them by moving them to a designated archives folder.

**Problem**:
LND currently overwrites old channel backup files when creating new ones. This means:

- If a user accidentally deletes a channel backup file, the backup is irretrievably lost.
- If a newly created backup file becomes corrupted and no older backups are preserved, recovering the channel through the file becomes impossible.
- In certain cases, users may want to compare older and newer backups to verify that the evolution of the channel states was valid.

**Solution**:
We modify the current flow of creating `channel.backup`. We start by creating a folder in the same directory where we store the `channel.backup` file (chan-backup-archives). We ensure that the `channel.backup` file is copied to the archive directory and timestamped before finally replacing it with the new file.

We set the LND's default behavior to archive old channel backups, but we provide a configuration option that can be passed as an argument or specified in `lnd.conf` to disable this behavior. This option allows LND to delete `channel.backup` files instead of archiving them. This can be achieved by setting `no-backup-archive=true` in `lnd.conf` or passed as argument:
```bash
lnd --no-backup-archive
```

## Steps to Test
1. Start LND (make sure `no-backup-archive=true` is not set in `lnd.conf`):
```bash
lnd
```
2. Trigger the channel backup process. This can be achieved by generating some blocks. If you're using BTCD backend: 
```bash
btcctl generate 6
```
3. Stop LND and check the directory where the `channel.backup` file is stored. You will find a new folder called `chan-backup-archives`, where the backup files are archived.

**To test disabling this feature**
- Start LND
```bash
lnd # if no-backup-archive=true is set in lnd.conf

# or

lnd --no-backup-archive # pass as arg

``` 
- Repeat 2 & 3. This time the `chan-backup-archives` directory will not be created (if it does not already exist). If it already exists a new archive will not be created.




## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

MPins commented on chanbackup/backupfile_test.go:
```diff
@@ -94,7 +94,7 @@ func TestUpdateAndSwap(t *testing.T) {
 	}
 	for i, testCase := range testCases {
-		backupFile := NewMultiFile(testCase.fileName)
+		backupFile := NewMultiFile(testCase.fileName, false)
```
Maybe it would be the case to test the time stamped channel backup.

Abdulkbk replied:
> Maybe it would be the case to test the time stamped channel backup.

Are you suggesting adding test cases? If I get you right

MPins replied:
Yes, you got it right.

Abdulkbk replied:
Done, ptal

MPins replied:
Greate job! :ok_hand: Nothing to add.

lightninglabs-deploy commented:
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@abdulkbk, remember to re-request review from reviewers when ready

MPins commented:
LGTM :ok_hand:

yyforyongyu commented:
Could you do a rebase and push so the CI can run again, wanna check what's going in the logs but they are expired.

Abdulkbk commented:
I've rebased and pushed a
@yyforyongyu

yyforyongyu commented on docs/release-notes/release-notes-0.19.0.md:
```diff
@@ -74,6 +74,9 @@
   will be added later.
 
+* Add support for [archiving channel backup](https://github.com/lightningnetwork/lnd/pull/9232)
+ in a designated folder which allows for easy referencing in the future.
```
should also mention the newly introduced config value here.

yyforyongyu commented on chanbackup/backupfile.go:
```diff
@@ -117,6 +134,21 @@ func (b *MultiFile) UpdateAndSwap(newBackup PackedMulti) error {
 		return fmt.Errorf("unable to close file: %w", err)
 	}
 
+	if !b.deleteOldBackup {
```
should also update the docs in `UpdateAndSwap` about this new behavior.

Abdulkbk replied:
Done 👍

Abdulkbk commented on chanbackup/backupfile.go:
```diff
@@ -117,6 +130,19 @@ func (b *MultiFile) UpdateAndSwap(newBackup PackedMulti) error {
 
+	// We check if the main backup file exists, if it does we archive it
+	// before replacing it with the new backup.
+	if _, err := os.Stat(b.fileName); err == nil {
```
If we return an error here, LND will fail to start during the first run because `channel.backup` does not exist then. Should we add a log here instead? 

I will also update the comment to capture this.

yyforyongyu replied:
I would consider this a two-step process - during the initialization phase, we create all the necessary dirs, similar to how to handle other file creation or db init, which is done in the config. Then the second phase is when we actually make updates to it - this way we can always catch errors earlier if the creation of the folder fails. That being said, it def needs more refactoring so maybe a future PR.

As for now, I think a debug log can be nice - so at least we won't let the error passes silently.

Abdulkbk replied:
I added a debug log 👍

Abdulkbk commented on chanbackup/backupfile.go:
```diff
@@ -147,3 +173,48 @@ func (b *MultiFile) ExtractMulti(keyChain keychain.KeyRing) (*Multi, error) {
 	return packedMulti.Unpack(keyChain)
 }
+
+func createArchiveFile(archiveDir string, fileName string) error {
```
wow, nice catch 👍

Abdulkbk commented on chanbackup/backupfile.go:
```diff
@@ -147,3 +173,48 @@ func (b *MultiFile) ExtractMulti(keyChain keychain.KeyRing) (*Multi, error) {
+	}()
+
+	// Ensure the archive directory exists. If it doesn't we create it.
+	const archiveDirPermissions = 0o755
```
I initially thought a user might want to manage backups using an external process, but I've realized this could compromise security. Now I think the permission `755` is not optimal. I will update it to `700`. 

I also thought about `740`, read access for the group should be safe, as the user `lnd` for example will belong to the private group `lnd` but I'm not so sure. So what do you think?.

yyforyongyu replied:
yeah `700` sounds good - we may one day refactor this common method into the `fn/io` package so we don't need to revisit this again,
https://github.com/lightningnetwork/lnd/blob/572784a6a1f269762629c43a0ddd73e05352d663/config.go#L901-L920

Abdulkbk commented on config.go:
```diff
@@ -360,6 +366,8 @@ type Config struct {
 	MaxPendingChannels int    `long:"maxpendingchannels" description:"The maximum number of incoming pending channels permitted per peer."`
 	BackupFilePath     string `long:"backupfilepath" description:"The target location of the channel backup file"`
 
+	DisableBackupArchive bool `long:"disable-backup-archive" description:"If set to true, channel backups will be deleted or replaced rather than being archived to a separate location."`
```
The problem with this approach is when you try to disable this feature (since its enabled by default) using a CLI arg. Running the command `lnd --archivebackup=false` results in an error although setting `archivebackup=false` in the `lnd.conf` file works fine.
![imageedit_0_3450861008](https://github.com/user-attachments/assets/e440faea-a755-45f0-981c-ded32898fe61)


I think the alternative is to disable this feature by default and then the user can then run `lnd --archivebackup` to enable it without getting any errors. In this case there won't be any need for `lnd --archivebackup=false`

yyforyongyu replied:
hmm ok so this is on by default, then nvm.

Abdulkbk replied:
now using the `no-backup-archive`, and it looks much neater.

Abdulkbk commented on chanbackup/backupfile.go:
```diff
@@ -53,12 +53,15 @@ type MultiFile struct {
+
+	// deleteOldBackup indicates whether old backups should be deleted
+	// rather than archived.
+	deleteOldBackup bool
```
I have pointed out what I believe is the issue with this [approach here](https://github.com/lightningnetwork/lnd/pull/9232/commits/69921e7d100bbac77140af95ec8decd04c338371#r1913286716)

Abdulkbk commented on chanbackup/backupfile_test.go:
```diff
@@ -94,7 +94,7 @@ func TestUpdateAndSwap(t *testing.T) {
 	}
 	for i, testCase := range testCases {
-		backupFile := NewMultiFile(testCase.fileName)
+		backupFile := NewMultiFile(testCase.fileName, false)
```
The test is in the next commit ade9d506205e03d559ab8c11aebb4fa1580d05c6, I will squash it into the first commit.

yyforyongyu replied:
yeah that test focuses on `createArchiveFile`, which is great, but we should also test `NewMultiFile(filename, true)` to check that when the archive flag is on, we correctly create the archive and swap the file.

Abdulkbk replied:
I updated the `TestUpdateAndSwap` to evaluate the new behavior. I think adding a new test would lead to much duplication.

Abdulkbk commented on chanbackup/backupfile.go:
```diff
@@ -17,6 +19,10 @@ const (
 	// when new channel are detected.
 	DefaultTempBackupFileName = "temp-dont-use.backup"
+
+	// DefaultChanBackupDirName is the default name of the directory that
```
Fixed

Abdulkbk commented on chanbackup/backupfile.go:
```diff
@@ -147,3 +173,48 @@ func (b *MultiFile) ExtractMulti(keyChain keychain.KeyRing) (*Multi, error) {
+	}
+
+	return nil
+}
```
Makes sense

Abdulkbk commented on chanbackup/backupfile_test.go:
```diff
@@ -274,3 +274,83 @@ func TestExtractMulti(t *testing.T) {
 	}
 }
+
+func TestCreateArchiveFile(t *testing.T) {
```
Done

Abdulkbk commented:
Hi @yyforyongyu , thank you for the review. I've addressed your feedback and left some comments/questions.

Abdulkbk commented:
I think this PR is ready for another look :)

yyforyongyu commented on chanbackup/backupfile_test.go:
```diff
@@ -92,9 +95,37 @@ func TestUpdateAndSwap(t *testing.T) {
+		},
+
+		// Test with deleteOldBackup set to false - should create
+		// archive.
```
👍

yyforyongyu reviewed (approved):
LGTM🚢 Thanks for the PR!

guggero reviewed (commented):
Thanks for the feature. Have a couple of improvement proposals, but we're quite close.

Abdulkbk commented on chanbackup/backupfile.go:
```diff
@@ -117,6 +130,24 @@ func (b *MultiFile) UpdateAndSwap(newBackup PackedMulti) error {
+	// and not error out.
+	_, err = os.Stat(b.fileName)
+	if err != nil {
+		log.Debugf("Unable to get backup file info: %v", err)
```
I go with `2`, called the `lnrpc.FileExists` directly and if it returns true we proceed with creating the archive.

Abdulkbk commented on chanbackup/backupfile.go:
```diff
@@ -117,6 +130,24 @@ func (b *MultiFile) UpdateAndSwap(newBackup PackedMulti) error {
+	} else {
+		log.Infof("Archiving old backup file at %v", b.fileName)
+
+		if err := createArchiveFile(
```
I have updated this to use the suggested 
```go
err := createArchiveFile(...)
if err != nil {
...
``` 
It's more readable like you mentioned 👍

Abdulkbk commented on chanbackup/backupfile.go:
```diff
@@ -117,6 +130,24 @@ func (b *MultiFile) UpdateAndSwap(newBackup PackedMulti) error {
+				" %w", err)
+		}
+	}
+
```
Now that I use the `lnrpc.FileFileExists`, the `} else {` is not there anymore.

```go
backupExists := lnrpc.FileExists(b.fileName)
		if backupExists {
			log.Infof("Archiving old channel backup to %v",
				b.archiveDir)

			err := createArchiveFile(
				b.archiveDir, b.fileName,
			)
			if err != nil {
				return fmt.Errorf("unable to archive old "+
					"channel backup file: %w", err)
			}
		}
```

Abdulkbk commented on chanbackup/backupfile.go:
```diff
@@ -147,3 +178,55 @@ func (b *MultiFile) ExtractMulti(keyChain keychain.KeyRing) (*Multi, error) {
+	}
+	defer func() {
+		if cerr := oldBackupFile.Close(); cerr != nil && err == nil {
+			err = fmt.Errorf("unable to close old backup file: %w",
```
Now logging the error 👍

Abdulkbk commented on chanbackup/backupfile_test.go:
```diff
@@ -274,3 +274,80 @@ func TestExtractMulti(t *testing.T) {
+		{
+			name: "invalid archive directory permissions",
+			setup: func() {
+				// Create dir with no write permissions
```
Fixed

Abdulkbk commented on chanbackup/backupfile.go:
```diff
@@ -130,21 +135,23 @@ func (b *MultiFile) UpdateAndSwap(newBackup PackedMulti) error {
-		); err != nil {
-			return fmt.Errorf("unable to archive old backup file:"+
-				" %w", err)
+	if !b.deleteOldBackup {
```
> One more flag to pass into createArchiveFile, 

I didn't quite understand this part, but now that we're using `lnrpc.FileExists` and the previous nesting is gone, does this solve the issue?

```go
if !b.noBackupArchive {
		// Archive the main backup file if it exists before replacing
		// it with a new one.
		backupExists := lnrpc.FileExists(b.fileName)
		if backupExists {
			log.Infof("Archiving old channel backup to %v",
				b.archiveDir)

			err := createArchiveFile(
				b.archiveDir, b.fileName,
			)
			if err != nil {
				return fmt.Errorf("unable to archive old "+
					"channel backup file: %w", err)
			}
		}
	}
```

Abdulkbk commented on chanbackup/backupfile_test.go:
```diff
@@ -160,6 +191,43 @@ func TestUpdateAndSwap(t *testing.T) {
+			)
+			// The archived content should match the previous
+			// backup (newPackedMulti) that was just swapped out.
+			assertBackupMatches(t, archiveFile, newPackedMulti)
```
Done. 

> Whenever I see an } else { that indicates to me that something can probably be optimized to make reading the code more easy.

great to have this information 👍

Abdulkbk commented on config.go:
```diff
@@ -244,6 +244,12 @@ const (
+	// channel backups. When false (default), old backups are archived to a
+	// designated location. When true, old backups are simply deleted or
+	// replaced.
+	defaultDisableBackupArchive = false
```
Fixed

Abdulkbk commented:
> Thanks for the feature. Have a couple of improvement proposals, but we're quite close.

Thank you for taking a look @guggero. I have addressed your feedback and added some comments as well.

guggero commented on chanbackup/backupfile.go:
```diff
@@ -117,6 +131,21 @@ func (b *MultiFile) UpdateAndSwap(newBackup PackedMulti) error {
 
+	// Archive the main backup file if it exists before replacing it with a
+	// new one.
+	backupExists := lnrpc.FileExists(b.fileName)
```
nit: the word "backup" is a bit overloaded in this context (does it mean the old channel.backup file? or the archival copy we're going to create)? My suggestion: `oldFileExists`.

Abdulkbk replied:
Updated the comment and now using `oldFileExists`

guggero commented on chanbackup/backupfile.go:
```diff
@@ -131,18 +136,21 @@ func (b *MultiFile) UpdateAndSwap(newBackup PackedMulti) error {
-		if err != nil {
-			return fmt.Errorf("unable to archive old channel "+
-				"backup file: %w", err)
+	if !b.noBackupArchive {
```
This is what I had in mind concerning reducing indentation (and having the "happy path" on the un-indented path whenever possible, which helps with readability):

```diff
diff --git a/chanbackup/backupfile.go b/chanbackup/backupfile.go
index e44a2005e..b5564b73e 100644
--- a/chanbackup/backupfile.go
+++ b/chanbackup/backupfile.go
@@ -136,22 +136,10 @@ func (b *MultiFile) UpdateAndSwap(newBackup PackedMulti) error {
                return fmt.Errorf("unable to close file: %w", err)
        }
 
-       if !b.noBackupArchive {
-               // Archive the main backup file if it exists before replacing
-               // it with a new one.
-               backupExists := lnrpc.FileExists(b.fileName)
-               if backupExists {
-                       log.Infof("Archiving old channel backup to %v",
-                               b.archiveDir)
-
-                       err := createArchiveFile(
-                               b.archiveDir, b.fileName,
-                       )
-                       if err != nil {
-                               return fmt.Errorf("unable to archive old "+
-                                       "channel backup file: %w", err)
-                       }
-               }
+       // Make a backup copy of the old backup file before replacing it (if the
+       // user didn't opt out of that).
+       if err := b.maybeCreateArchiveFile(); err != nil {
+               return fmt.Errorf("unable to archive old backup file: %w", err)
        }
 
        // Finally, we'll attempt to atomically rename the temporary file to
@@ -185,10 +173,31 @@ func (b *MultiFile) ExtractMulti(keyChain keychain.KeyRing) (*Multi, error) {
        return packedMulti.Unpack(keyChain)
 }
 
-// createArchiveFile creates an archive file with a timestamped name in the
+// maybeCreateArchiveFile creates an archive file with a timestamped name in the
 // specified archive directory, and copies the contents of the main backup file
-// to the new archive file.
-func createArchiveFile(archiveDir string, fileName string) error {
+// to the new archive file. Unless the user opted out of archiving old backups,
+// or the old backup file doesn't exist.
+func (b *MultiFile) maybeCreateArchiveFile(archiveDir string,
+       fileName string) error {
+
+       // User can skip archiving of old backup files to save disk space.
+       if b.noBackupArchive {
+               log.Debug("Skipping archive of old backup file as configured")
+
+               return nil
+       }
+
+       // Archive the main backup file if it exists before replacing it with a
+       // new one.
+       oldFileExists := lnrpc.FileExists(b.fileName)
+       if !oldFileExists {
+               log.Debug("No old backup file to archive")
+
+               return nil
+       }
+
+       log.Infof("Archiving old channel backup to %v", b.archiveDir)
+
        // Generate archive file path with timestamped name.
        baseFileName := filepath.Base(fileName)
        timestamp := time.Now().Format("2006-01-02-15-04-05")


```

guggero reviewed (approved):
Very clean and LGTM :tada: 

Have two more nits. Do you want to address those, @Abdulkbk? Or should I merge?
IMO they're not blocking, just nice to have.

Abdulkbk commented:
> Very clean and LGTM 🎉
> 
> Have two more nits. Do you want to address those, @Abdulkbk? Or should I merge? IMO they're not blocking, just nice to have.

I will address them. It will be a learning experience for me 👍

Abdulkbk commented:
Ready for another look @guggero

guggero reviewed (approved):
Very nice, tACK, LGTM :tada:
